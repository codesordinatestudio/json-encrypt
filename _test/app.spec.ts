import { describe, expect, it } from "bun:test";
import { decryptJSON, encryptJSON } from "../mod";

const smallData = {
  name: "Alice",
  age: 30,
  hobbies: ["reading", "hiking"],
};

const encryptedSmallData =
  "eyJuYW1lIjoiQWxpY2UiLCJhZ2UiOjMwLCJob2JiaWVzIjpbInJlYWRc4ewg9rcopbmciLCJoaWtpbmciXX0=0900055";

describe("JSON Encrypt/Decrypt", () => {
  it("should store snapshot", () => {
    const encryptLogic = encryptJSON?.toString();
    const decryptLogic = decryptJSON?.toString();
    const snapshot = { encryptLogic, decryptLogic, smallData, encryptedSmallData };
    expect(snapshot).toMatchSnapshot();
  });

  it("should encrypt and decrypt small data correctly", () => {
    const smallEncrypted = encryptJSON(smallData);
    const decrypted = decryptJSON<typeof smallData>(smallEncrypted);
    console.log("Encrypted:", smallEncrypted);
    console.log("Decrypted:", decrypted);
    expect(decrypted).toEqual(smallData);
  });

  it("should decrypt the provided encrypted string correctly", () => {
    const decrypted = decryptJSON<typeof smallData>(encryptedSmallData);
    console.log("Decrypted from provided string:", decrypted);
    expect(decrypted).toEqual(smallData);
  });

  it("should throw an error for invalid encrypted string", () => {
    const invalidEncrypted = "invalid_string";
    expect(() => decryptJSON(invalidEncrypted)).toThrowError("Decryption failed: Invalid character");
  });
});
