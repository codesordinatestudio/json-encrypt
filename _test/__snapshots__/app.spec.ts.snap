// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Encrypt/Decrypt should store snapshot 1`] = `
{
  "decryptLogic": 
"function decryptJSON(encrypted) {
  try {
    const insertPositionStr = encrypted.slice(-5);
    const keyLengthStr = encrypted.slice(-7, -5);
    const insertPosition = parseInt(insertPositionStr, 10);
    const keyLength = parseInt(keyLengthStr, 10);
    const encodedWithKey = encrypted.slice(0, -7);
    const base64 = encodedWithKey.substring(0, insertPosition) + encodedWithKey.substring(insertPosition + keyLength);
    const jsonStr = atob(base64);
    return JSON.parse(jsonStr);
  } catch (error) {
    throw new Error(\`Decryption failed: Invalid character\`);
  }
}"
,
  "encryptLogic": 
"function encryptJSON(json) {
  const jsonStr = JSON.stringify(json);
  const base64 = btoa(jsonStr);
  const keyLength = Math.floor(Math.random() * 5) + 8;
  const key = Math.random().toString(36).substring(2, 2 + keyLength);
  const insertPosition = Math.floor(Math.random() * base64.length);
  const encodedWithKey = base64.substring(0, insertPosition) + key + base64.substring(insertPosition);
  const keyLengthStr = keyLength.toString().padStart(2, "0");
  const insertPositionStr = insertPosition.toString().padStart(5, "0");
  return encodedWithKey + keyLengthStr + insertPositionStr;
}"
,
  "encryptedSmallData": "eyJuYW1lIjoiQWxpY2UiLCJhZ2UiOjMwLCJob2JiaWVzIjpbInJlYWRc4ewg9rcopbmciLCJoaWtpbmciXX0=0900055",
  "smallData": {
    "age": 30,
    "hobbies": [
      "reading",
      "hiking",
    ],
    "name": "Alice",
  },
}
`;
